File: Ch05/learning_tf/src/turtle_tf_listener.cpp
#include <ros/ros.h>
#include <tf/transform_listener.h>
#include <geometry_msgs/Twist.h>
#include <turtlesim/Spawn.h>

int main(int argc, char** argv){
  ros::init(argc, argv, "my_tf_listener");  // 初始化节点

  ros::NodeHandle node; // 创建节点句柄

  ros::service::waitForService("spawn");  // 等待名为spawn的服务
  // 创建一个service client，尖括号中是服务的类型
  ros::ServiceClient add_turtle =
    node.serviceClient<turtlesim::Spawn>("spawn");
  // 创建该类型的srv
  turtlesim::Spawn srv;
  // 调用服务，存到srv
  add_turtle.call(srv);

  ros::Publisher turtle_vel =
    node.advertise<geometry_msgs::Twist>("turtle2/cmd_vel", 10);

  // 创建一个TransformListener对象,一旦监听器被创建，它开始接收tf转换，并缓冲它们长达10秒
  tf::TransformListener listener;

  ros::Rate rate(10.0);
  while (node.ok()){
    tf::StampedTransform transform;
    // 从/turtle2坐标系开始变换到/turtle1坐标系,提供ros::Time(0)即会给出最近的可用的变换,结果存放的变换对象transform
    try{
      listener.lookupTransform("/turtle2", "/turtle1",
                               ros::Time(0), transform);
    }
    catch (tf::TransformException &ex) {
      ROS_ERROR("%s",ex.what());
      ros::Duration(1.0).sleep();
      continue;
    }

    // 变换用于计算小海龟的新的线性和角速度，基于它与龟的距离和角度
    // 新的速度发布在话题"turtle2/cmd_vel"中，turtlesim将使用它来更新turtle2的运动
    geometry_msgs::Twist vel_msg;
    vel_msg.angular.z = 4.0 * atan2(transform.getOrigin().y(),
                                    transform.getOrigin().x());
    vel_msg.linear.x = 0.5 * sqrt(pow(transform.getOrigin().x(), 2) +
                                  pow(transform.getOrigin().y(), 2));
    turtle_vel.publish(vel_msg);

    rate.sleep();
  }
  return 0;
};

File: Ch05/learning_tf/scripts/turtle_tf_listener.py
#!/usr/bin/env python
#coding:utf-8

import roslib
roslib.load_manifest('learning_tf')
import rospy
import math
import tf
import geometry_msgs.msg
import turtlesim.srv

if __name__ == '__main__':
	# 初始化节点
    rospy.init_node('turtle_tf_listener')

    # 创建一个tf.TransformListener对象，用于接收TF转换并缓冲10秒（TF包提供了一个tf.transformlistener实现使接收任务将更容易）
    listener = tf.TransformListener()

    rospy.wait_for_service('spawn')
    spawner = rospy.ServiceProxy('spawn', turtlesim.srv.Spawn)
    spawner(4, 2, 0, 'turtle2')

    turtle_vel = rospy.Publisher('turtle2/cmd_vel', geometry_msgs.msg.Twist,queue_size=1)

    rate = rospy.Rate(10.0)
    while not rospy.is_shutdown():
        try:
        	# 通过listener.lookupTransform函数来实现获取变换
        	# 第一参数：变换从/turtle2坐标系，第二参数：到/turtle1坐标系
        	# rospy.Time(0)，想变换的时间，rospy.Time(0)获取最新的变换
            (trans,rot) = listener.lookupTransform('/turtle2', '/turtle1', rospy.Time(0))
        except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
            continue

        # 设置参数来控制跟随小海龟（turtle2）的运动
        angular = 4 * math.atan2(trans[1], trans[0])
        linear = 0.5 * math.sqrt(trans[0] ** 2 + trans[1] ** 2)
        cmd = geometry_msgs.msg.Twist()
        cmd.linear.x = linear
        cmd.angular.z = angular
        turtle_vel.publish(cmd)

        rate.sleep()







